### PR TITLE
implement the sender secretbox

### PR DESCRIPTION
**Pull request by oconnor663**
@maxtaco 

This saves us 37 bytes per-recipient-after-the-first, (32 for the repeated sender key, and 5 more in overhead since we no longer encrypt a MessagePack structure), at cost of 48 bytes in the new encrypted sender field. In theory we could chop off another 16 by omitting the Poly1305 tag for the sender's key (because the payload will fail to decrypt if it's modified), but we're avoiding the lower-level NaCl functions. Not a big deal.

### CODE REVIEW

File: go/saltpack/decrypt.go [hunk 1]
```diff
@@ -18,8 +18,8 @@ type decryptStream struct {
 	mps        *msgpackStream
 	err        error
 	state      readState
-	keys       *receiverKeysPlaintext
-	sessionKey SymmetricKey
+	payloadKey SymmetricKey
```
Comment (maxtaco) [inline go/saltpack/decrypt.go]: can this be a `*SymmetricKey*`?

**Comment by oconnor663** [comment]
Security-wise, here's what's changed:

- **Before** this change we were encrypting the sender's public key with the ephemeral private key. That meant that only someone in possession of that key could lie about it. If recipients other than the original sender wanted to lie, they needed to generate a new ephemeral keypair and a whole new fake header, which broke payload authenticators in _two_ ways (wrong sender key _and_ wrong nonce).
- **After** this change we encrypt the sender's public key with the payload key. That means that any recipient can modify the sender key as though they were the original sender. Now this only breaks payload authenticators in _one_ way (wrong sender key).

If I have that right, this doesn't matter to us. We already have to detect the original sender lying about the sender key. Allowing other recipients to try the same attack doesn't change that. If sender anonymity weren't an issue, we could write the sender key in the clear and still not be worried.

File: go/saltpack/decrypt.go (hunk 1 above)
Comment (oconnor663) [inline go/saltpack/decrypt.go]: `payloadKey` is just a renaming of `sessionKey` to match the spec. The type hasn't changed. If you prefer the term "session key" though I can totally keep it.

Comment (maxtaco) [inline go/saltpack/decrypt.go]: Meaning, can this be a pointer?

File: go/saltpack/decrypt.go [hunk 2]
```diff
@@ -145,26 +145,27 @@ func (ds *decryptStream) tryVisibleReceivers(hdr *EncryptionHeader, ephemeralKey
...
+		return nil, nil, SymmetricKey{}, -1, ErrBadLookup
 	}
 
-	keysRaw, err := shared.Unbox(ds.nonce.ForKeyBox(), hdr.Receivers[orig].Keys)
+	payloadKeySlice, err := shared.Unbox(ds.nonce.ForKeyBox(), hdr.Receivers[orig].PayloadKeyBox)
 	if err != nil {
-		return nil, nil, nil, -1, err
+		return nil, nil, SymmetricKey{}, -1, err
 	}
+	var payloadKey SymmetricKey
```
Comment (maxtaco) [inline go/saltpack/decrypt.go]: can you make a construct that loads from a slice?  it should potentially error out too

File: go/saltpack/decrypt.go [hunk 3]
```diff
@@ -179,15 +180,17 @@ func (ds *decryptStream) tryHiddenReceivers(hdr *EncryptionHeader, ephemeralKey
 
 		for i, r := range hdr.Receivers {
 			if len(r.ReceiverKID) == 0 {
-				keysRaw, err := shared.Unbox(ds.nonce.ForKeyBox(), r.Keys)
+				payloadKeySlice, err := shared.Unbox(ds.nonce.ForKeyBox(), r.PayloadKeyBox)
 				if err == nil {
-					return secretKey, shared, keysRaw, i
+					var payloadKey SymmetricKey
```
Comment (maxtaco) [inline go/saltpack/decrypt.go]: can only use this fabled constructor here too

File: go/saltpack/decrypt.go (hunk 2 above)
Comment (oconnor663) [inline go/saltpack/decrypt.go]: Good idea, we can merge that with `verifyRawKey`.

**Comment by oconnor663** [comment]
Coverage restored to previous!

**Comment by maxtaco** [comment]
:+1: great job!

**Pull request closed without merging**